### ISSUE TITLE
Dealing with AssertionErrors

### ISSUE THREAD

**Issue by khand100** [ZombieAlex/MFCAuto#21]
Hi,
sometimes MFCAuto throws AssertionErrors which are coming from the Model merge function:
```
AssertionError [ERR_ASSERTION]: Merging a non-model? Non-models need some special casing that is not currently implemented.
    at Model.merge (G:\Downloads\XChats\node_modules\MFCAuto\lib\Model.js:98:16)
    at rdata.forEach (G:\Downloads\XChats\node_modules\MFCAuto\lib\Client.js:152:59)
    at Array.forEach (<anonymous>)
    at Client._packetReceived (G:\Downloads\XChats\node_modules\MFCAuto\lib\Client.js:148:39)
    at Client._readPacket (G:\Downloads\XChats\node_modules\MFCAuto\lib\Client.js:276:18)
    at Client._readData (G:\Downloads\XChats\node_modules\MFCAuto\lib\Client.js:49:14)
    at Socket.client.on (G:\Downloads\XChats\node_modules\MFCAuto\lib\Client.js:634:34)
    at emitOne (events.js:115:13)
    at Socket.emit (events.js:210:7)
    at addChunk (_stream_readable.js:266:12)
```

Sadly they crash my whole application, right know I have removed the assert check in the merge function and instead just let it return. Is there a better way to handle/ignore theses?

Because it looks like this only happens when a specific user is online or in a model room.
During my debugging I changed the error message to get the uid which is causing this.
It is 5666491 which resolves to 'FightRight' with a lv of 2 (Premium), and the assert violation happens multiple times per second. Maybe he is spamming malformed packets or something (the reason for this assumption is the Anonymous logo in his profile pic :D )

Edit: I am still using version 4.2.2, but I saw the same code is still in v5

**Comment by ZombieAlex** [comment]
Hmm, interesting. Removing the assert should be fine for now.

The assert is there because I can't guarantee proper state or event callbacks for members. If you're not depending upon those then you should be fine until I can think of a better fix.

Sometimes MFC will send a packet about a user and not say what level they are. So in some places, if it's the first time I'm seeing any information about a user and the server hasn't indicated whether the user is a member or not, I err on the side of assuming it's a model so that I don't miss any model state updates.

This is the sequence:

1. MFC tells us about him but doesn't indicate his level
2. We err on the "safe" side and add him to the Model collection
3. MFC later tell us more about him and *does* include his level
4. We find a Model instance for him and try to merge it
5. Merge does that check and asserts

Possible real fixes are to:
- Simply remove the assert like you've done. But that would lead to slow memory leaks in the Model state as we accumulate members that never get flagged as offline
- Change to assume that users with no levels are members instead of models. But I don't know if that would cause us to lose any model status updates or not.
- In merge, when we finally detect that a "model" is actually a member, silently remove him from the global Model state at that point.

Probably the third option is the best. I'll think about it a bit more.

Thanks for the report.

**Comment by khand100** [comment]
Thanks for the fast reply,
with my limited understanding of MFCAuto I agree with you, the 3rd fix looks like the best.

Until then one could also use the uncaughtException handler, if one dont want to edit library files. But I am not quit sure if this has side effects. Because like I said before, when this happens the error gets triggered multiple times a second, at least in my application (listening to nearly all mfc models).

```
process.on("uncaughtException", err => {
    if(err.name.includes("AssertionError")){
        // ignore or log
    }else{
        throw err;
    }
});
```

**Comment by ZombieAlex** [comment]
I implemented the third fix and pushed v5.1.0. More broadly I would not advise swallowing assertion errors as you proposed. Assertions *always* represent bugs. In this case, the bug was relatively safe to ignore, in other cases it may not be. Please continue to report assertions to me if you hit them.

Thanks!

**Comment by khand100** [comment]
Thanks now I only have to upgrade to v5, but this should not be a problem.
